### PR TITLE
web: codemod `<Element className="btn" />`occurrences

### PR DESCRIPTION
**Pull request by valerybugakov**
## Context

Semi-automated migration of the elements using `className="btn"` to [the `<Button />` Wildcard component](https://github.com/sourcegraph/sourcegraph/blob/main/client/wildcard/src/components/Button/Button.tsx) powered by [the `buttonElementToComponent` codemod](https://github.com/sourcegraph/codemod/tree/main/packages/transforms/src/buttonElementToComponent).

### Notes

- Components based on `shared` components like `WebHoverOverlay` will be updated in the follow-up PR after merging [the `Alert` migration PR](https://github.com/sourcegraph/sourcegraph/pull/29862). 
- @courier-new, please check out changes to the `ButtonTooltip` component: https://github.com/sourcegraph/sourcegraph/commit/6d3b785b2273db3a207f7eab9215a82544921791 and the `InputTooltip` component: https://github.com/sourcegraph/sourcegraph/commit/4b2f17d073e157b1b797c855bc4ed4ce2bdb224a.
- `* migration` commits are automated migrations made by codemod.

## Outstanding work

`SearchOnboardingTour`:

https://github.com/sourcegraph/sourcegraph/blob/e1369056695f5cf064e9f089b9505e0e043b64bb/client/web/src/search/input/SearchOnboardingTour.ts#L75-L79

@novoselrok, I didn't look into this one closely yet, but right away curious why raw DOM manipulations are used here instead of regular React components? What are there potential blockers for the migration to React components?

Generated by the codemod:

```tsx
./client/web/src/components/shared.tsx:11:110 - warning: Class 'btn btn-link' is used on the 'CommandListPopoverButton' element. Please update it manually.
>>>    
    <CommandListPopoverButton
        {...props}
        buttonClassName="btn btn-link"
        popoverClassName="popover  border-0"
        popoverInnerClassName="rounded overflow-hidden"
        formClassName="form p-2 bg-1 border-bottom"
        inputClassName="form-control px-2 py-1"
        listClassName="list-group list-group-flush list-unstyled pt-1"
        actionItemClassName="list-group-item list-group-item-action p-2 border-0"
        selectedActionItemClassName="active border-primary"
        noResultsClassName="list-group-item text-muted"
    />

./client/web/src/nav/GlobalNavbar.tsx:276:29 - warning: Class 'btn btn-link p-0 m-0' is used on the 'WebCommandListPopoverButton' element. Please update it manually.
>>>    <WebCommandListPopoverButton
                                {...props}
                                location={location}
                                buttonClassName="btn btn-link p-0 m-0"
                                menu={ContributableMenu.CommandPalette}
                                keyboardShortcutForShow={KEYBOARD_SHORTCUT_SHOW_COMMAND_PALETTE}
                            />

./client/web/src/components/WebHoverOverlay/WebHoverOverlay.tsx:103:13 - warning: Class 'btn btn-sm btn-secondary border-0' is used on the 'HoverOverlay' element. Please update it manually.
>>>    
        <HoverOverlay
            {...propsToUse}
            className={classNames('card', styles.webHoverOverlay)}
            actionItemClassName="btn btn-sm btn-secondary border-0"
            onAlertDismissed={onAlertDismissed}
            getAlertClassName={getAlertClassName}
        />

./client/web/src/repo/docs/RepositoryDocumentationPage.tsx:197:29 - warning: Class 'mr-1 text-decoration-none btn-link' is used on the 'Button' element. Please update it manually.
>>>    <Button
                                target="_blank"
                                rel="noopener"
                                href="https://docs.sourcegraph.com/code_intelligence/apidocs"
                                className="mr-1 text-decoration-none btn-link"
                                variant="secondary"
                                outline={true}
                                size="sm"
                                as="a"
                            >

./client/web/src/search/notebook/SearchNotebookPageHeaderActions.tsx:91:21 - warning: Class 'btn-danger' is used on the 'DropdownItem' element. Please update it manually.
>>>    <DropdownItem className="btn-danger" onClick={() => setShowDeleteModal(true)}>
```

Part of https://github.com/sourcegraph/sourcegraph/issues/29318.
Related to https://github.com/sourcegraph/codemod/issues/31.

### CODE REVIEW

File: client/wildcard/src/types.ts [hunk 1]
```diff
@@ -17,18 +17,21 @@ export interface ForwardReferenceComponent<
...
      * We explicitly avoid `React.ElementType` and manually narrow the prop types
      * so that events are typed when using JSX.IntrinsicElements.
      */
-    <As = IntrinsicElementString>(
+    <As = IntrinsicElementString, OwnPropsOverwrite = OwnProps>(
         props: As extends ''
             ? { as: keyof JSX.IntrinsicElements }
             : As extends React.ComponentType<infer P>
-            ? Merge<P, OwnProps & { as: As }>
+            ? Merge<P, Omit<OwnPropsOverwrite, 'ref'> & { as: As }>
```
Comment (vovakulikov) [inline client/wildcard/src/types.ts:32]: Is this a case when a Component has a custom prop that is called `ref` in its props? Do we have this in our codebase?

File: client/web/src/components/ButtonTooltip.tsx [hunk 2]
```diff
@@ -11,16 +15,9 @@ type ButtonProps = React.DetailedHTMLProps<React.ButtonHTMLAttributes<HTMLButton
  *
  * All other props are passed to the `button` element.
  */
-export const ButtonTooltip: React.FunctionComponent<
-    ButtonProps & { tooltip?: string } & Required<Pick<ButtonProps, 'type'>>
-> = ({ disabled, tooltip, type, ...props }) => (
+export const ButtonTooltip: React.FunctionComponent<ButtonTooltipProps> = ({ disabled, tooltip, ...props }) => (
```
Comment (courier-new) [inline client/web/src/components/ButtonTooltip.tsx:18]: Cool timing on this one actually -- I just opened https://github.com/sourcegraph/sourcegraph/pull/30075 on Friday to actually bring this tooltip-restoring behavior back into `wildcard/Button` and get rid of this component entirely. 🙂 Replaces the need for the changes here, but the changes here also look fine on their own.

File: client/web/src/enterprise/batches/detail/changesets/BatchChangeChangesetsHeader.tsx [hunk 3]
```diff
@@ -18,7 +18,7 @@ export const BatchChangeChangesetsHeader: React.FunctionComponent<BatchChangeCha
         {toggleSelectAll && (
             <InputTooltip
                 type="checkbox"
-                className="btn ml-2"
+                className="ml-2"
```
Comment (courier-new) [inline client/web/src/enterprise/batches/detail/changesets/BatchChangeChangesetsHeader.tsx:21]: Most (if not all) instances of `InputTooltip` are `type="checkbox"`, but I think this attribute will get stripped out of the `Button` because of [this check](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@vb/button-migration/-/blob/client/wildcard/src/components/Button/Button.tsx#L62:24):

```tsx
type={Component === 'button' ? type : undefined}
```

This causes the `InputTooltip`s to be rendered as text input fields instead of checkboxes, like this instance:

image

These look normal again if we always allow `type` to be set on `Button`:

```tsx
type={type}
```

(or just omitting the line and including it with `{...attributes}`)

Comment (valerybugakov) [reply client/web/src/enterprise/batches/detail/changesets/BatchChangeChangesetsHeader.tsx:21]: Thanks for looking into it! I should've delayed tagging people in the PR description until fixing all the outstanding issues. After addressing them, I plan to move the PR out of the draft stage.

File: client/web/src/components/InputTooltip.tsx [hunk 4]
```diff
@@ -1,19 +1,31 @@
...
-> = ({ disabled, tooltip, ...props }) => (
+export const InputTooltip: React.FunctionComponent<InputTooltipProps> = ({ disabled, tooltip, type, ...props }) => (
     <div className={styles.container}>
         {disabled ? <div className={styles.disabledTooltip} data-tooltip={tooltip} /> : null}
-        <input disabled={disabled} data-tooltip={disabled ? undefined : tooltip} {...props} />
+        <Button
+            as="input"
+            disabled={disabled}
+            data-tooltip={disabled ? undefined : tooltip}
+            type={type as ButtonProps['type']}
```
Comment (valerybugakov) [inline client/web/src/components/InputTooltip.tsx:27]: We discussed this issue with @vovakulikov, and typecasting will be redundant after addressing this issue: https://github.com/sourcegraph/sourcegraph/issues/30348.
Decided to do it properly in a separate PR.

File: client/web/src/components/LoaderButton.tsx [hunk 5]
```diff
@@ -1,9 +1,9 @@
 import classNames from 'classnames'
 import React from 'react'
 
-import { LoadingSpinner } from '@sourcegraph/wildcard'
+import { LoadingSpinner, Button, ButtonProps } from '@sourcegraph/wildcard'
 
-interface Props extends React.DetailedHTMLProps<React.ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement> {
+interface Props extends Omit<ButtonProps, 'as'> {
```
Comment (umpox) [inline client/web/src/components/LoaderButton.tsx:6]:
```suggestion
interface Props extends ButtonProps {
```

**Review by umpox — APPROVED** [review]
👍

File: client/wildcard/src/components/Button/Button.tsx [hunk 6]
```diff
@@ -6,9 +6,7 @@ import { ForwardReferenceComponent } from '../../types'
 import { BUTTON_VARIANTS, BUTTON_SIZES } from './constants'
 import { getButtonSize, getButtonStyle } from './utils'
 
-export interface ButtonProps
-    extends React.ButtonHTMLAttributes<HTMLButtonElement>,
-        React.RefAttributes<HTMLButtonElement> {
+export interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
```
[the remark this answers is not in the document]
Comment (valerybugakov) [reply client/wildcard/src/components/Button/Button.tsx:9]: Sadly, it causes issues in cases where we extend `ButtonProps`. HTML props are not included, and it causes a lot of errors. We discussed it with @vovakulikov, as I mentioned in the comment [here](https://github.com/sourcegraph/sourcegraph/pull/30096#discussion_r794483718). It will be addressed in the follow-up.

File: client/search-ui/src/results/NoResultsPage.tsx [hunk 7]
```diff
@@ -49,18 +49,20 @@ const SearchInputExample: React.FunctionComponent<SearchInputExampleProps> = ({
                 {showSearchContext && (
                     <>
                         <div className={classNames(searchBoxStyle.searchBoxContextDropdown, styles.fakeSearchContext)}>
-                            <div
+                            <Button
+                                as="div"
```
Comment (vovakulikov) [inline client/search-ui/src/results/NoResultsPage.tsx:53]: I understand that this is probably is out of this PR scope but Button as div looks a little bit suspicious. I mean if it looks like a button and it works like a button it should be a button.

File: client/web/src/components/InputTooltip.tsx [hunk 8]
```diff
@@ -1,19 +1,31 @@
...
  */
-export const InputTooltip: React.FunctionComponent<
-    React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement> & { tooltip: string }
-> = ({ disabled, tooltip, ...props }) => (
+export const InputTooltip: React.FunctionComponent<InputTooltipProps> = ({ disabled, tooltip, type, ...props }) => (
     <div className={styles.container}>
         {disabled ? <div className={styles.disabledTooltip} data-tooltip={tooltip} /> : null}
-        <input disabled={disabled} data-tooltip={disabled ? undefined : tooltip} {...props} />
+        <Button
+            as="input"
```
Comment (vovakulikov) [inline client/web/src/components/InputTooltip.tsx:24]: I know that this is also kind of out of scope but I'm really interested in this case, Button as input? @courier-new whenever you have time can you share what is the case for input like button?

File: client/web/src/components/LinkWithIcon.tsx [hunk 9]
```diff
@@ -1,44 +1,31 @@
...
-        to,
-        className: classNames('d-flex', 'align-items-center', className),
-        ...(activeClassName && { activeClassName }),
-        ...omit(restProps, ['hasIconPlaceholder', 'icon']),
-    }
+export const LinkWithIcon: React.FunctionComponent<LinkWithIconProps> = props => {
+    const { to, exact, text, className, activeClassName, icon: Icon } = props
 
     return (
-        <LinkComponent {...linkProps} data-testid={kebabCase(text)}>
```
Comment (vovakulikov) [inline client/web/src/components/LinkWithIcon.tsx:39]: @valerybugakov don't we need this `data-testid` anymore?

File: client/web/src/components/Sidebar.tsx [hunk 10]
```diff
@@ -18,25 +18,21 @@ export const SidebarNavItem: React.FunctionComponent<{
     exact?: boolean
     source?: string
 }> = ({ children, className, to, exact, source }) => {
-    const buttonClassNames = classNames('btn text-left d-flex', styles.linkInactive)
+    const buttonClassNames = classNames('text-left d-flex', styles.linkInactive, className)
+    const routeMatch = useRouteMatch(to)
 
     if (source === 'server') {
         return (
-            <a href={to} className={classNames(buttonClassNames, className)}>
```
Comment (vovakulikov) [inline client/web/src/components/Sidebar.tsx:25]: @valerybugakov can we include in the migration roadmap migration for `a` → `Link`.

File: client/web/src/enterprise/code-monitoring/components/CodeMonitorForm.module.scss [hunk 11]
```diff
@@ -4,10 +4,7 @@
     margin: calc(1rem - 1px) 0;
 
     &-link {
-        &:hover {
-            text-decoration: none;
-            color: var(--link-color);
-        }
+        color: var(--link-color);
```
Comment (vovakulikov) [inline client/web/src/enterprise/code-monitoring/components/CodeMonitorForm.module.scss:7]: @valerybugakov do we need this? I thought that the btn-link class does this styling. Also, I thought it doesn't do anything around `text-decoration: none;` is it safe to delete this?

File: client/web/src/enterprise/insights/pages/dashboards/dashboard-page/components/dashboard-menu/DashboardMenu.tsx [hunk 12]
```diff
@@ -35,10 +37,11 @@ export const DashboardMenu: React.FunctionComponent<DashboardMenuProps> = props
     return (
         <Menu>
             <MenuButton
+                as={Button}
                 ref={innerRef}
                 data-tooltip={tooltipText}
                 data-placement="right"
-                className={classNames(styles.triggerButton, 'btn btn-icon')}
+                className={classNames(styles.triggerButton, 'btn-icon')}
```
Comment (vovakulikov) [inline client/web/src/enterprise/insights/pages/dashboards/dashboard-page/components/dashboard-menu/DashboardMenu.tsx:44]: @valerybugakov will we handle `btn-icon` in a separate PR? Just curious why didn't we include the btn-icon as a possible variant for the `variant` prop?

File: client/web/src/enterprise/insights/pages/dashboards/dashboard-page/components/dashboard-menu/DashboardMenu.tsx [hunk 13]
```diff
@@ -35,10 +37,11 @@ export const DashboardMenu: React.FunctionComponent<DashboardMenuProps> = props
     return (
         <Menu>
             <MenuButton
+                as={Button}
```
Comment (vovakulikov) [inline client/web/src/enterprise/insights/pages/dashboards/dashboard-page/components/dashboard-menu/DashboardMenu.tsx:40]: @umpox I thought we got rid of the reach-ui `MenuButton` component. Is it something that has been added after we did the menu-button migration?  Or I'm wrong and we just implemented `MenuButton` component but haven't started adoption?

**Review by vovakulikov — APPROVED** [review]
@valerybugakov cool work. I left a few questions but they are not blockers

File: client/web/src/enterprise/insights/pages/dashboards/dashboard-page/components/dashboard-menu/DashboardMenu.tsx (hunk 13 above)
Comment (umpox) [reply client/web/src/enterprise/insights/pages/dashboards/dashboard-page/components/dashboard-menu/DashboardMenu.tsx:40]: The migration is still in progress by GitStart: https://github.com/sourcegraph/sourcegraph/issues/27701

It looks like they're still focusing on migrating away from the `Reactstrap` components first

File: client/search-ui/src/results/NoResultsPage.tsx (hunk 7 above)
Comment (umpox) [reply client/search-ui/src/results/NoResultsPage.tsx:53]: Yeah I thought this too, I think it's OK for the migration PR - but as part of the accessibility work I wonder if we should narrow the `as` type for this component - maybe just allow links

File: client/web/src/components/LinkWithIcon.tsx (hunk 9 above)
Comment (valerybugakov) [reply client/web/src/components/LinkWithIcon.tsx:39]: Thanks for noticing this: https://github.com/sourcegraph/sourcegraph/pull/30357

File: client/search-ui/src/results/NoResultsPage.tsx (hunk 7 above)
Comment (valerybugakov) [reply client/search-ui/src/results/NoResultsPage.tsx:53]: After pulling component implementations into the `wildcard`, we will go over their usage and create issues for such use-cases.

File: client/web/src/components/Sidebar.tsx (hunk 10 above)
Comment (valerybugakov) [reply client/web/src/components/Sidebar.tsx:25]: Yes, it's the part of [the `Link` migration](https://github.com/sourcegraph/sourcegraph/issues/27663).

File: client/web/src/components/InputTooltip.tsx (hunk 8 above)
Comment (courier-new) [reply client/web/src/components/InputTooltip.tsx:24]: Yeah I think all instances of this can and should be migrated to `Checkbox`es. The only reason I'm not racing to do that is the default `BaseControlInput` styles don't work well in the instances of these, because they happen to want the default styles that are closer to the `Button` instead. 😅 So they're really not buttons, they just happen to all want `className="btn"` right now.

File: client/web/src/enterprise/insights/pages/dashboards/dashboard-page/components/dashboard-menu/DashboardMenu.tsx (hunk 12 above)
Comment (valerybugakov) [reply client/web/src/enterprise/insights/pages/dashboards/dashboard-page/components/dashboard-menu/DashboardMenu.tsx:44]: It wasn't added to the initial implementation of the `Button` component. We're in the migration process now, and soon it should disappear from the codebase.
